Bump deps to latest major version and other misc dev UX

fire332 opened the pull request:
# Changes
- Bump all dependencies to their latest **major** version. I have reviewed the changes for all the dependencies and they either don't break anything we do or I've fixed it. Things to note:
  - Most major version bumps are due to dropping End-Of-Life'ed Node versions.
  - `babel-plugin-polyfill-corejs3`'s breaking change is described eloquently here: https://github.com/babel/babel/issues/15663#issuecomment-1564849962
- Added more ESLint rules that weren't enabled by the recommended config. No linting errors were introduced.
- Cleaned up tooling configs by removing options that effectively do nothing.
- Fix deploy script broken by #128.
- Tooling configs are now type checked and linted.
- Removed deprecated VSCode extension from recommendation.
- Switched to ESLint flat config.
- Add `type-check` script and added all required `@types` to `devDependencies`
- Added `"type": "module` to `package.json`. This means `.js` files ran by Node will be interpreted as ESM by default.
- `lint-staged` now runs ESLint and tsc as well.

throwaway96 commented on tools/deploy.mjs:
```diff
@@ -4,7 +4,7 @@ process.exit(
     'node',
     [
-      './node_modules/@webosose/ares-cli/bin/ares-install.js',
+      './node_modules/@webos-tools/cli/bin/ares-install.js',
```
Should we really be reaching into `node_modules` like that? It seems likely to break in exactly this way. I would prefer [libnpmexec](https://github.com/npm/cli/tree/latest/workspaces/libnpmexec) or at least `./node_modules/.bin/ares-install`.

fire332 replied:
If you're fine with `shell: true`, the latter is easily done.

throwaway96 replied:
Is that necessary for Windows? (Does it still use `.cmd` files on Windows?) On Linux, just replacing `node` with `./node_modules/.bin/ares-install` and removing the script path from the args seems to work.

fire332 replied:
It's necessary on Windows. To quote Node docs:

> On Windows, however, .bat and .cmd files are not executable on their own without a terminal, and therefore cannot be launched using child_process.execFile().

https://nodejs.org/api/child_process.html#spawning-bat-and-cmd-files-on-windows

throwaway96 replied:
Then it's fine. Does everything else work on Windows?

throwaway96 commented:
Just a couple of tangentially-related notes:

I would like to move to the ESLint flat config format before our configs get too complex. Since it's currently mostly just rules, porting shouldn't be too bad. But I want to use `eslint-plugin-prettier`, and once you start with plugins it can get messy.

Does `ares-install` work reliably for you? I've had to come up a custom deploy script, since `ares-install` fails like 80% of the time for me (and I'm not the only one).

The lack of trailing commas drives me a bit nuts (merge conflicts...), but I'm not going to change it at least until there are fewer pending PRs.

fire332 commented:
> Does everything else work on Windows?

I normally run `npm run build && npm run package && npm run deploy` every time I want to view my changed so it better work 😅.

> Does ares-install work reliably for you?

Yes, I've went through 3 versions of `ares-cli` and it's my main method of installing.

> I would like to move to the ESLint flat config format before our configs get too complex.

I don't have any experience with flat configs because it was only a few weeks ago that `typescript-eslint` started supporting flat configs and I had been busy since that update came out 😕.

fire332 commented:
> I want to use `eslint-plugin-prettier`

Any particular reason why? I find that having squigglies in the editor for formatting issues just get in the way so I usually just setup my dev workflows to have auto formatting be as behind-the-scenes as possible.

In addition, mixing prettier issues with ESLint means that if you ever want to make automated PR code reviews from ESLint running in GH workflows, prettier errors will bleed in and spew dozens of garbage reviews when a simple "this PR needs to be formatted with prettier" status check would suffice.

throwaway96 commented:
I like to see what changes formatters are making, including so that I'm less likely to be surprised by them mangling code. The ESLint plugin seems to be the [recommended](https://github.com/prettier/prettier/issues/6069#issuecomment-484363765) way to accomplish that.

In my experience the output for automated checks on PRs isn't that bad, but I suppose it could be worked around if it becomes an issue. We have automatic linting of PRs with `eslint-plugin-prettier` on Homebrew Channel, and it seems to work pretty well.

throwaway96 reviewed (changes requested):
Maybe I'm doing something wrong, but none of the userscript stuff seems to be working and I'm getting this in the console:

```
Uncaught ReferenceError: require is not defined
    at Module../src/domrect-polyfill.js (<anonymous>:14652:13)
    at __webpack_require__ (<anonymous>:17442:42)
    at <anonymous>:17548:75
    at <anonymous>:17605:2
    at <anonymous>:17606:12
```

This is on webOS 6.3.2 (K7Lp).

I tried production and development builds of the PR both as-is and merged onto v0.3.3-rc2. Looks like there are a bunch of `require()`s, but I don't have time to look into it any further right now.

fire332 commented:
> Maybe I'm doing something wrong, but none of the userscript stuff seems to be working and I'm getting this in the console:
> 
> ```
> Uncaught ReferenceError: require is not defined
>     at Module../src/domrect-polyfill.js (<anonymous>:14652:13)
>     at __webpack_require__ (<anonymous>:17442:42)
>     at <anonymous>:17548:75
>     at <anonymous>:17605:2
>     at <anonymous>:17606:12
> ```
> 
> This is on webOS 6.3.2 (K7Lp).
> 
> I tried production and development builds of the PR both as-is and merged onto v0.3.3-rc2. Looks like there are a bunch of `require()`s, but I don't have time to look into it any further right now.

FIxed. Babel isn't smart enough to determine whether an import is ESM or CommonJS based on their `package.json` and Webpack isn't smart enough to handle that anyway. So I just added empty `export` statements to the bottom of two files to force Babel to treat those as ESM.

throwaway96 commented:
This seems to fix the webpack production build on webOS 3.0 (Chromium 38), which was previously causing `Uncaught SyntaxError: Unexpected token ,` (#151). Any idea why? Was something relevant changed here, or is it just a coincidence?

fire332 commented:
Babel bumped 7.24.3 -> 7.24.4 ([CHANGELOG](https://github.com/babel/babel/releases/tag/v7.24.4)). Please confirm the `Uncaught SyntaxError` issue remains resolved.

throwaway96 commented:
That seems to be fine. It does run, but the menu is broken. However, I can't really test webOS 3.0 properly, so I can't tell whether these issues are genuine or caused by my setup.

The (hosted, at least) DevTools for Chromium 38 don't show a preview, and newer DevTools just cause it to crash. VNC screws up WAM (and webOS) in general in various ways. Most of my webOS 3.0 boards are K2L(p), which seems to have lots of problems when not connected to a proper t-con/panel. I have one M2 board, but it runs SimpleSmart, not webOS. My only M16 boards are broken, commercial, or both (plus an STB).

So unless someone wants to buy me an M2 or M16 consumer webOS board with Vx1 or LVDS output, I can't properly test anything on webOS 3.0.

fire332 commented:
Are you using a different GitHub app? I've responded to all of them a week ago.

throwaway96 commented:
No, I'm using the GitHub site. I don't see any response.
![image](https://github.com/webosbrew/youtube-webos/assets/68320646/b4b92443-786f-45e9-a4dd-effdbfd8a253)

fire332 commented:
![Screenshot of all of fire332's replies](https://github.com/webosbrew/youtube-webos/assets/96039230/403aa2df-8b37-467d-b659-94720913fbde)

throwaway96 commented:
Huh. I guess the reason I don't see them is that they're "Pending", although I'm not sure what that means exactly.

And yes, I think enabling `prefer-character-class` would be a good idea.

fire332 commented:
...

Quality GitHub UX design.

throwaway96 commented on src/index.html:
```diff
@@ -1,4 +1,4 @@
-<!DOCTYPE html>
+<!doctype html>
```
I don't really like it, but whatever.

throwaway96 commented on webpack.config.js:
```diff
@@ -1,34 +1,31 @@
       rules: [
         {
-          test: /\.m?js$/,
+          test: /\.(?:m|c)?js$/i,
```
Yes.

fire332 commented:
Done. Do you want me configure lint-staged to run ESLint and tsc as well?

throwaway96 commented:
Not in this PR. I'd like to get a new release done ASAP.

fire332 commented:
If time is the issue, I already have the config done for a previous project. I just need to paste it in.

throwaway96 commented:
I guess if you can do it in like the next 30 minutes.

fire332 commented:
Done.

throwaway96 commented on lint-staged.config.js:
```diff
@@ -0,0 +1,10 @@
+function globForCode(/** @type {string} */ toolName) {
+  return '*.?(c|m){js,ts}?(x)' + `*(${toolName})`;
```
What's the point of this? To have multiple tasks for the same effective pattern?

Sorry, I'm not familiar with `lint-staged` configuration.

fire332 replied:
Yes. If it's done this way, the tasks are ran in parallel.

throwaway96 replied:
And you want them to be separate because you're passing filenames to one (`eslint`) but not the other (`tsc`), so you can't use an array?

throwaway96 replied:
All right. Looks weird, but whatever.

fire332 replied:
An array of tasks are ran in serial. We can use an array but that means no parallel execution for speedup.

fire332 commented:
Oops. I just realized, I forgot to add `lint-staged.config.js` to `tsconfig.tooling.json`

throwaway96 commented:
Just ran into this error while testing. No idea whether it's related to this PR.

```
VM22:2 Uncaught (in promise) TypeError: Kt is not a function
    at t.<anonymous> (<anonymous>:2:157100)
    at S (<anonymous>:2:110448)
    at Generator.<anonymous> (<anonymous>:2:111764)
    at Generator.next (<anonymous>:2:110851)
    at h (<anonymous>:2:116298)
    at a (<anonymous>:2:116489)
    at <anonymous>:2:116548
    at new Promise (<anonymous>)
    at new e (<anonymous>:2:31826)
    at t.<anonymous> (<anonymous>:2:116435)
    at t.<anonymous> (<anonymous>:2:157957)
    at <anonymous>:2:163908
```
    
It's on the line with `return n = Kt(this.videoID).substring(0, 4),`.
```js
e = [{
key: "init",
value: (r = v(d().mark((function t() {
    var e, r, n, o, i, a, s, c = this;
    return d().wrap((function(t) {
        for (; ; )
            switch (t.prev = t.next) {
            case 0:
                return n = Kt(this.videoID).substring(0, 4),
                o = ["sponsor", "intro", "outro", "interaction", "selfpromo", "music_offtopic"],
                t.next = 4,
                fetch(Et(e = Et(r = "".concat("https://sponsorblock.inf.re/api", "/skipSegments/")).call(r, n, "?categories=")).call(e, encodeURIComponent(Nt(o))));
```

Is it possible to manually specify a source map for a userscript (it seems to always appear as `VM22`)? I haven't been able to figure out a way.

fire332 commented:
> Is it possible to manually specify a source map for a userscript (it seems to always appear as VM22)? I haven't been able to figure out a way.

We can swap from `devtool: 'source-map'` to `inline-source-map`.

throwaway96 commented:
> We can swap from `devtool: 'source-map'` to `inline-source-map`.

Thanks, that does work. But I'm still not sure how `tiny-sha256` broke.

fire332 commented:
Does webpack strip source maps from production bundles? Have you tested the dev bundle? Sorry, I'll be unavailable for the next half hour or so.

throwaway96 commented:
Changing
```js
import * as sha256 from 'tiny-sha256';
```
to
```js
import sha256 from 'tiny-sha256';
```
fixes the issue. But I don't actually know JavaScript, so I have no idea why.

fire332 commented:
The former collects all exports into an object while the latter only uses the default export for ESM. For CommonJS interop, most bundlers will convert `module.exports` to the default export and `module.exports.anything` into a named export. But sometimes a package has a bad `package.json` or the bundler has some broken logic which makes "* as something" work.

throwaway96 commented:
So some change to webpack/babel/whatever in this PR means the latter form is now required?

fire332 commented:
Yes. It's standard practice anyways. The former is a holdover from when ESM was new and bundler devs didn't really know how interop should work.

throwaway96 commented:
I think it's working, but I can't really test it because the SponsorBlock server seems to be down.
